cancat/uds.py

```diff
@@ -8,39 +8,61 @@ def __init__(self, c, tx_arbid, rx_arbid):
         self.c = c
         self.tx_arbid = tx_arbid
         self.rx_arbid = rx_arbid
-        self.t = threading.Thread(target = self.SendTesterPresent)
-        self.TesterPresent = False
-        self.t.setDaemon(True)
-        self.t.start()
 
-    def ReassembleIsoTP(self, start_index = 0):
+    def ReassembleIsoTP(self, start_index = 0, service = None):
         msg_found = False
         while msg_found is False:
             for idx, ts, arbid, msg in self.c.genCanMsgs(start=start_index, arbids=[self.rx_arbid]):
-                msg_found = True
-                return hex(arbid), msg.encode('hex')
+                # Check that the message is for the expected service, if specified
+                if service is not None:
+                    # Check if this is the right service, or there was an error
+                    if ord(msg[1]) == service or ord(msg[1]) == 0x7f:
+                        msg_found = True
+                        return arbid, msg
+                else:
+                    msg_found = True
+                    return arbid, msg
             time.sleep(0.1)
 
     def SendTesterPresent(self):
-        while True:
-            if self.TesterPresent:
-                self.c.CANxmit(self.tx_arbid, "023E000000000000".decode('hex'))
+        while self.TesterPresent is True:
+            self.c.CANxmit(self.tx_arbid, "023E000000000000".decode('hex'))
             time.sleep(2.0)
 
     def StartTesterPresent(self):
         self.TesterPresent = True
+        self.t = threading.Thread(target = self.SendTesterPresent)
```

Comment:
The problem is that the object will never be ready to be trashed since the thread will always hold a reference to the object. The threads I'm creating should be garbage collected, so creating a bunch of them in the same session shouldn't be an issue. I'll grant there's some overhead to thread spinup but this function isn't going to be called frequently. Also, there are a lot of instances where we may use UDS functionality without sending the TesterPresent command, so we save the thread overhead completely in those cases.